src/chain_table.rs

```diff
@@ -94,33 +95,41 @@ impl<const TABLE_WIDTH: usize, const BLOCK_LENGTH: usize>
     ///
     /// - `rng`         : random number generator
     /// - `kwi_value`   : DEM key used to encrypt the value
-    pub fn encrypt<const KEY_LENGTH: usize, DEM: Dem<KEY_LENGTH>>(
+    pub fn encrypt(
         &self,
         rng: &mut impl CryptoRngCore,
-        kwi_value: &DEM::Key,
+        kwi_value: &Aes256GcmKey,
     ) -> Result<Vec<u8>, Error> {
-        let bytes = self.try_to_bytes()?;
-        DEM::encrypt(rng, kwi_value, &bytes, None).map_err(Error::from)
+        let nonce = Nonce::new(rng);
+        let plaintext = self.serialize()?;
+        let mut result = Vec::with_capacity(
+            nonce.as_bytes().len() + plaintext.len() + Aes256Gcm::ENCRYPTION_OVERHEAD,
+        );
+        result.extend(nonce.as_bytes());
+        result.extend(
+            Aes256Gcm::new(kwi_value)
+                .encrypt(&nonce, &plaintext, None)
+                .map_err(Error::from)?,
+        );
+        Ok(result)
     }
 
     /// Decrypts the Chain Table value using the given DEM key.
     ///
     /// - `kwi_value`   : DEM key used to encrypt the value
     /// - `ciphertext`  : encrypted Chain Table value
-    pub fn decrypt<const DEM_KEY_LENGTH: usize, DEM: Dem<DEM_KEY_LENGTH>>(
-        kwi_value: &DEM::Key,
-        ciphertext: &[u8],
-    ) -> Result<Self, Error> {
-        let max_ciphertext_length = 1 + TABLE_WIDTH * (1 + BLOCK_LENGTH) + DEM::ENCRYPTION_OVERHEAD;
+    pub fn decrypt(kwi_value: &Aes256GcmKey, ciphertext: &[u8]) -> Result<Self, Error> {
+        let max_ciphertext_length =
+            1 + TABLE_WIDTH * (1 + BLOCK_LENGTH) + Aes256Gcm::ENCRYPTION_OVERHEAD;
         if max_ciphertext_length != ciphertext.len() {
             return Err(Error::CryptoError(format!(
                 "invalid ciphertext length: given {}, should be {}",
                 ciphertext.len(),
                 max_ciphertext_length
             )));
         }
-        let bytes = DEM::decrypt(kwi_value, ciphertext, None)?;
-        Self::try_from_bytes(&bytes)
+        let (nonce, ct) = ciphertext.split_at(Aes256Gcm::NONCE_LENGTH);
+        Self::deserialize(&Aes256Gcm::new(kwi_value).decrypt(&Nonce::try_from(nonce)?, ct, None)?)
```

Comment:
This function panics in case the ciphertext is invalid (`if ciphertext.len() < Aes256Gcm::NONCE_LENGTH`). You could get a slice of it instead:
```suggestion
        let nonce = Nonce::try_from_slice(&ciphertext[..Aes256Gcm::NONCE_LENGTH])?;
        Self::deserialize(&Aes256Gcm::new(kwi_value).decrypt(
            &nonce,
            &ciphertext[Aes256Gcm::NONCE_LENGTH..],
            None,
        )?)
```

Reply:
done